lib/util/selector-parser.js

```diff
@@ -0,0 +1,255 @@
+        return consume(index);
+    }
+
+    /**
+     * Parse key
+     * @returns {string} key of the attribute
+     */
+    function parseKey() {
+        return consume(index);
+    }
+
+    /**
+     * Parse query
+     * @returns {Object} Query
+     */
+    function parseQuery() {
+        var token = tokens[index];
+        var query = [];
+        while (token !== "=") {
+            if (token !== ".") {
+                query.push(parseKey());
+            } else {
+                token = consume(index);
+            }
+            token = tokens[index];
+        }
+        return query;
+    }
+
+    /**
+     * Parse attribute
+     * @returns {Object} Attribute
+     */
+    function parseAttribute() {
+        var query = parseQuery();
+        consume(index);
+        var value = parseValue();
+        return {
+            query: query,
+            value: value
+        };
+    }
+
+    /**
+     * Parse attributes
+     * @returns {Object[]} Array of attributes
+     */
+    function parseAttributes() {
+        var token = tokens[index],
+            attributes = [];
+        while (token !== "]" || tokens[index + 1] === "[") {
+            if (token === "]") {
+                consume();
+                consume();
+            }
+            attributes.push(parseAttribute());
+            token = tokens[index];
+        }
+        return attributes;
+    }
+
+    /**
+     * Parse type
+     * @returns {Object} Node type
+     */
+    function parseType() {
+        return consume(index);
+    }
+
+    /**
+     * Parse node
+     * @returns {Object} Node
+     */
+    function parseNode() {
+        var type,
+            attributes,
+            node = {};
+        if (tokens[index] === "[") {
+            consume(index);
+            attributes = parseAttributes();
+        } else {
+            type = parseType();
+            if (tokens[index] === "[") {
+                consume(index);
+                attributes = parseAttributes();
+            }
+        }
+        if (tokens[index] === "]") {
+            consume(index);
+        }
+
+        if (type) {
+            node.type = type;
+        }
+        if (attributes) {
+            node.attributes = attributes;
+        }
+        return node;
+    }
+
+    /**
+     * Parse relationship
+     * @param {Object} parent - Parent node
+     * @returns {Object} node
+     */
+    function parseRelationship(parent) {
+        var relationshipType = consume(index),
+            decendant = parseNode();
+        if (relationshipType === ">") {
+            parent.hasDirectChild = true;
+            decendant.directChild = true;
+        }
+        return decendant;
+    }
+
+    /**
+     * Parse selector
+     * @returns {Object} Selector
+     */
+    function parseSelector() {
+        var selector = {
+                event: item,
+                selectors: []
+            },
+            node = parseNode();
+        selector.selectors.push(node);
```

Comment:
Agree:-) I got stuck at naming this one, and decided to go with whatever first came to my mind. Top level variable should probably be changed to something else (query?)